go-controller/pkg/clustermanager/userdefinednetwork/controller_helper.go

```diff
@@ -43,51 +48,58 @@ func (c *Controller) updateNAD(obj client.Object, nad *netv1.NetworkAttachmentDe
 		}
 
 		// TODO: add lock for primary NAD creation, avoid conflict with other primary UDN creators
-		nad, err := c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(obj.GetNamespace()).Create(context.Background(), desiredNAD, metav1.CreateOptions{})
+		newNAD, err := c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(namespace).Create(context.Background(), desiredNAD, metav1.CreateOptions{})
 		if err != nil {
 			return nil, fmt.Errorf("failed to create NetworkAttachmentDefinition: %w", err)
 		}
-		klog.Infof("Created NetworkAttachmentDefinition [%s/%s]", nad.Namespace, nad.Name)
+		klog.Infof("Created NetworkAttachmentDefinition [%s/%s]", newNAD.Namespace, newNAD.Name)
 
-		return nad, nil
+		return newNAD, nil
 	}
 
-	if !metav1.IsControlledBy(nad, obj) {
-		return nil, fmt.Errorf("foreign NetworkAttachmentDefinition with the desired name already exist [%s/%s]", nad.Namespace, nad.Name)
+	if !metav1.IsControlledBy(nadCopy, obj) {
+		return nil, fmt.Errorf("foreign NetworkAttachmentDefinition with the desired name already exist [%s/%s]", nadCopy.Namespace, nadCopy.Name)
 	}
 
-	if !reflect.DeepEqual(nad.Spec.Config, desiredNAD.Spec.Config) {
-		nad.Spec.Config = desiredNAD.Spec.Config
-		uNAD, uerr := c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(nad.Namespace).Update(context.Background(), nad, metav1.UpdateOptions{})
-		if uerr != nil {
-			return nil, fmt.Errorf("failed to update NetworkAttachmentDefinition: %w", uerr)
-		}
-		klog.Infof("Updated NetworkAttachmentDefinition [%s/%s]", uNAD.Namespace, uNAD.Name)
-		return uNAD, nil
+	if reflect.DeepEqual(nadCopy.Spec.Config, desiredNAD.Spec.Config) {
+		return nadCopy, nil
+	}
+
+	nadCopy.Spec.Config = desiredNAD.Spec.Config
+	updatedNAD, err := c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(nadCopy.Namespace).Update(context.Background(), nadCopy, metav1.UpdateOptions{})
```

Comment:
so on update we allow to update the spec? I thought there were certain fields that were immutable for UDNs. Are those enforced in the CRD rather than in this code?